### DISCUSSION TITLE
ES9018 DAC

### DISCUSSION THREAD

**Discussion by alidaf** [sle118/squeezelite-esp32#395] [Q&A]
I've just replaced a PCM5102 DAC with an [ES9018](https://www.aliexpress.com/item/1005003740290057.html?spm=a2g0o.order_list.order_list_main.42.21ef1802rj35fU) but I get nothing but bad hiss in the left speaker. Has anyone had any success with the ES9018?

**Comment by sle118** [comment]
Could you How did you configure it? Do you have the data sheets? The DAC might need a MCLK to work properly (which is supported, of course).

**Reply by alidaf** [reply]
The ES9018 is on another level for DACs and there's plenty of documentation for the chip itself, but the problem is when certain 3rd parties release a board that is almost certainly an attempt at a knock-off of the Audiophonics boards for the RPi, which are woefully expensive. There is scant information other than a few screen shots of a configuration menu with the HifiBerry DAC selected. The are a couple of badly translated FAQs, one of which covers the MCLK, which isn't needed. I took a punt but fell short on this one.

**Comment by ElFishi** [comment]
I have a [ES9018K2M running](https://github.com/sle118/squeezelite-esp32/issues/40#issuecomment-1935478949 ) with squeezelite-esp32, even w/o MCLK.
The default input configuration of the ES9018K2M is I2S 32-bit, and it works with the 32bit FW.
To run it 16-bit you need to set the input register.
`dac_controlset {"init":[ {"reg":1,"val":12} ],"poweron":[ {"reg":1,"val":12} ]}`

**Reply by alidaf** [reply]
Hi again @ElFishi. I got so far with this after hijacking your other post because it's the only one that mentions the ES9018 with the ESP32 anywhere on the entire internet as far as I can tell. Obviously this is due to your singular and insane level of wizardry.
The problem is that this board doesn't have any user I2C access so the settings are hard-coded and obfuscated. The site claims that this works with the Moode and Volumio RPi distros and the HifiBerry DAC overlay, but that board is a PCM5102 so I don't get how they can make that claim. It should be a straightforward swap but the only progress I made was to use the 32-bit version of Squeezelite-ESP32. By progress, I mean I no longer get a repulsive hiss, just silence.

**Reply by martymcbro** [reply]
Hi,

I'm also trying to switch my ES9018K2M to 16bit to have access to the equalizer function. However the dac_controlset above doesn't seem to work for me. I can hear the music but it's distorted and a lot louder than in 32bit mode. I checked the spec of the ES9018K2M and came to the conclusion that the value 12 for register 1 should be correct. I entered the string above without the "dac_controlset" under dac_controlset in the nvs editor and rebooted. In the config backup it looks like "dac_controlset": "{\"init\":[ {\"reg\":1,\"val\":12} ],\"poweron\":[ {\"reg\":1,\"val\":12} ]}",
My board is this one (updated I2S input): https://a.aliexpress.com/_EI5tTRe

I connected the MCLK to GND.

Do you have an idea why it doesn't work for me?

**Reply by alidaf** [reply]
> Hi,
> 
> I'm also trying to switch my ES9018K2M to 16bit to have access to the equalizer function. However the dac_controlset above doesn't seem to work for me. I can hear the music but it's distorted and a lot louder than in 32bit mode. I checked the spec of the ES9018K2M and came to the conclusion that the value 12 for register 1 should be correct. I entered the string above without the "dac_controlset" under dac_controlset in the nvs editor and rebooted. In the config backup it looks like "dac_controlset": "{"init":[ {"reg":1,"val":12} ],"poweron":[ {"reg":1,"val":12} ]}", My board is this one (updated I2S input): https://a.aliexpress.com/_EI5tTRe
> 
> I connected the MCLK to GND.
> 
> Do you have an idea why it doesn't work for me?

That looks identical to my board but it doesn't have i2c so the dac_controlset won't be doing anything. Scroll down to the bottom for the config I used to get it working. Also, this only works in 32 bit mode.

**Reply by martymcbro** [reply]
My board is already working fine in 32bit mode with the output configured as I2S. So my board seems to be the I2S version. I'm only missing the equalizer function in LMS and I thought it might be because of the 32 bit version. On my other 16bit boards the equalizer is available. So I wanted to use the 16bit version and set a dac_controlset.

When I understand correctly what you wrote below is that there is a SPDIF version of the board which you could either solder to become the I2S version (what you did) or you can change the output of the ESP to spdif and change the wiring between the ESP and the DAC accordingly.

From my understanding the dac_controlset can only be set via I2S. You're running the 32bit version with I2S output. Are you able to use the 16bit version and set a dac_controlset? Or is the ES9018 always connected via SPDIF and the I2S input from the ESP is only converted to SPDIF by a seperate chip on the DAC board?

**Reply by alidaf** [reply]
The equaliser is only available in the 16bit version of Squeezelite-ESP32 as far as I can tell. Finding the detail in the guide is pretty difficult but I recall seeing that somewhere.

My board is the I2S input, DAC output version and does not have any I2C connections for configuring the bit depth etc, which is what I thought the dac_controlset uses. The two sound similar but I2C is a communication protocol and I2S is a format. Essentially my board is hardwired in 24bit depth mode and can't be changed because it doesn't have I2C, which is why I have to use the 32 bit version of Squeezelite-ESP32.

The only soldering I did was to re-solder the joints because they were pretty poor quality.

For SPDIF, I have to change the configuration in the Squeezelite-ESP32 interface but it is wired to a different GPIO than the DAC. It's not possible to have both at the same time from the ESP32 board directly. That said, I recently bought an SPDIF board that appears to have I2S passthrough, so I was hoping to try and wire the DAC to that and just have a single configuration. I think it depends on whether the DAC can use the MLCK from the SPDIF board because the SPDIF board re-clocks.

**Reply by martymcbro** [reply]
OK, that makes sense. Now I understand why the dac_controlset doesn't work. I haven't set the I2C connection and there are probably a few pins missing for that. So only 32 bit firmware with these boards.

Thank you for your help

**Reply by ElFishi** [reply]
![grafik](https://github.com/user-attachments/assets/02b4a972-de64-49b2-a760-ecd64e57d510)
You can connect the I2C pins SCL, SDA, if you find them on the board and are bold enough to wire some enameled wire to the pads. The Aliexpress pictures aren't very good, but I suspect the traces to the pins are under the S and 9 of the ES9018 silk screen label. Probably then pins are pulled up.

**Reply by martymcbro** [reply]
Ah, OK. Thank you for the hint. I'll keep that in mind. I think right now it's not so important for me.

**Comment by ElFishi** [comment]
I just put that down if any other poor soul finds this in google. 
My ES9018K2M plays proper music with the 32bit I2S firmware of squeezelite-esp32 and an empty dac_controlset, i.e. no command via I2C.
Are you sure your wiring is OK? 
I do not use MCLK, but [Sébastien](https://github.com/sle118) says you can.  
Do you have any schematic for your board?

**Reply by alidaf** [reply]
I thought as much but figured it would still be handy to at least indicate how far this has got here. I do still wonder whether the board is faulty or I've been sent the wrong version but there are obvious translation issues in dealing with the company.

**Comment by ElFishi** [comment]
I just looked at the photos of you board on Aliexpress. It looks like the I2C pins (3 SCL, 4 SDA) are connected to the PCB and lead to 10k resistors pulling them high. So with a steady hand you could solder two wires to access the I2C port. 
This will not solve your problem, though, just sayin'.

**Reply by alidaf** [reply]
It might not solve the problem, but what a hoot! I'm very tempted. I'm about to try removing a potentiometer on a valve amp board to replace it with a pair of digital pots so I can have IR and rotary encoder control. If I don't go up in smoke with that then I may give it a go because, damn it, I want to be a wizard too, even if I'm the oldest in the class!

**Reply by sle118** [reply]
That's the spirit!

**Comment by alidaf** [comment]
Update. I'm not going to go into what happened with the tube amp (until I've recovered), but as far as the ES9018K2M board is concerned, and after an awful lot of to-ing and fro-ing, with some mild rage (a little more than mild according to family members) and associated greying, I finally got it to work with Moode using the Audiophonics-ES9018 driver. The seller's latest excuse is that RaspiOS is to blame because the ES9018 driver no longer exists in Volumio. They seem oblivious to the fact that they advise using a driver for the PCM5102. But anyways!
The question remains as to whether this board could still work with the Squeezelite-ESP32. The RPi drivers are [here](https://github.com/luoyi/Rpi-ES9018K2M-DAC/tree/master).
There is no I2C access. The RPi connections are pure I2S, as they would be for the ESP32 so I don't know how the device gets configured. I think it's magic. There's a lot of it about lately.

**Comment by ElFishi** [comment]
If you can't send commands via I2C, feed it with 32 bits I2S. It should work.

**Reply by alidaf** [reply]
I don't know how to change the I2S bit length. I've looked all over, which probably means it's in bold headline and I just can't see it.

**Reply by ElFishi** [reply]
its in the choice of firmware 
![grafik](https://github.com/sle118/squeezelite-esp32/assets/29286528/e45d0471-6b3c-459f-80fa-8e3ace86050b)

**Reply by alidaf** [reply]
Ah, ok. Yes I'm already using 32 bit. I think the audio bit-depth is still only a maximum of 24 bits though. I don't know how the RPi does it but it must do some sort of resampling on the fly because I think the ES9018 is fixed to 32 bits.

**Reply by ElFishi** [reply]
Its only the interface configuration for the I2S input. It could be as simple as zero padded. But the default configuration of the ES9018K2M is that you toggle the serial clock 32 times and send 32 bits for one channel, then toggle the LR-clock, send another 32 bit, toggle LR again and so on. 
The 32-bit fw does that no matter what bit depth your audio source has.

**Reply by alidaf** [reply]
That is actually more information than I've managed to find in the last 3 weeks. That makes it sound trivial to implement but it would probably have to be done as a preset and hard-coded.

**Comment by dierochade** [comment]
I have very similar problems with this dac: [AUDIOPHONICS DAC Sabre ES9018K2M I2S / SPDIF to RCA 32bit 384kHz DSD256 ](https://www.audiophonics.fr/en/dac-modules/audiophonics-dac-sabre-es9018k2m-i2s-spdif-to-rca-32bit-384khz-dsd256-p-15975.html)

![Audiophonics dac9018k2m-1](https://github.com/sle118/squeezelite-esp32/assets/125600882/97381429-aa33-49fd-b394-d32a2f8d8e30)
![Audiophonics dac9018k2m-2](https://github.com/sle118/squeezelite-esp32/assets/125600882/41ad559c-cc0a-4d97-b67b-d9df4fdde757)

I run the normal 32bit firmware . 
Me too, I do only get some noise, farly resembling the music on one channel. So for me this does not work out of the box?

When aked for the wiring, I did just test it for continuity, which is ok. Soldering looks good for me too.
Is it correct that I cannot connect to i2c on this board, because there are no pins 

Is there anything that can be done to make this work?
Would it have any use to ask the seller for documentation (Its a french company, so they might give support. What would be useful resources, schematics, data sheet and ??

Any help or recommendation (change dac? Use spdif?) would be highly appreciated

**Reply by alidaf** [reply]
That's correct, no need for I2C. It is pre-configured.

That board is the similar to mine but has the extra 74HC04D inverter chip of the "coax" version as it's sold on the website I bought mine, although it still has the third capacitor that is normally removed for something else. I imagine it would work the same way and the connections should be the same apart from the inputs.

I don't know specifically what made mine work in the end but to be safe I re-soldered the connections on both the DAC board and the ESP32. I also made sure to take the 5V and ground directly from the ESP32. I tend to split the power supply before the ESP32 (but link grounds), to power peripherals rather than risk overloading the ESP32. This isn't usually a problem but could have made a difference here. I really don't know for sure.

In any case, I was thankfully prompted to try one more time by a resident wizard guru, re-connected everything as simply as I could and it worked straight away.

The only config for mine is in the NVS editor: `model=I2S,bck=25,ws=33,do=32`

**Reply by alidaf** [reply]
Obviously, the NVS parameter for SPDIF is different, i.e. spdif_config, which for the same GPIOs will just be `bck=25,ws=33,do=32`
